doc/malheur.man
.\" Automatically generated by Pod::Man 2.27 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "MALHEUR 1"
.TH MALHEUR 1 "2017-09-26" "Malheur 0.6.0" "User Manual"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
malheur \- Automatic analysis of malware behavior
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBmalheur\fR [\fBoptions\fR] \fIaction\fR \fIdataset\fR [\fIdataset\fR ...]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBmalheur\fR is a tool for the automatic analysis of malware behavior (program
behavior recorded from malicious software in a sandbox environment).  The
tool has been designed to support the regular analysis of malicious software
and the development of detection and defense measures.  \fBmalheur\fR allows
for identifying novel classes of malware with similar behavior and assigning
unknown malware to discovered classes.  It supports four basic actions for
analysis which can be applied to reports of recorded behavior:
.IP "\fIExtraction of prototypes\fR." 4
.IX Item "Extraction of prototypes."
From a given set of reports, \fBmalheur\fR identifies a subset of prototypes
representative for the full data set.  The prototypes provide a quick
overview of recorded behavior and can be used to guide manual inspection.
.IP "\fIClustering of behavior\fR." 4
.IX Item "Clustering of behavior."
\&\fBmalheur\fR automatically identifies groups (clusters) of reports containing
similar behavior.  Clustering allows for discovering novel classes of
malware and provides the basis for crafting specific detection and defense
mechanisms, such as anti-virus signatures.
.IP "\fIClassification of behavior\fR." 4
.IX Item "Classification of behavior."
Based on a set of previously clustered reports, \fBmalheur\fR is able to assign
unknown behavior to known groups of malware.  Classification enables
identifying novel and unknown variants of malware and can be used to filter
program behavior prior to manual inspection.
.IP "\fIIncremental analysis\fR." 4
.IX Item "Incremental analysis."
\&\fBmalheur\fR can be applied incrementally for analysis of large data
sets. By processing reports in chunks, the run-time as well as memory
requirements can be significantly reduced. This renders long-term
application of \fBmalheur\fR feasible, for example for daily analysis 
of incoming malware programs.
.PP
A detailed description of these techniques as well as technical 
background on analysis of malicious software is provided in the 
following articles:
.IP "\(bu" 4
\&\fIAutomatic Analysis of Malware Behavior using Machine Learning\fR.
Konrad Rieck, Philipp Trinius, Carsten Willems, and Thorsten Holz.
Journal of Computer Security (\s-1JCS\s0), 19(4) 639\-668, 2011.
.IP "\(bu" 4
\&\fIA Malware Instruction Set for Behavior-Based Analysis\fR.
Philipp Trinius, Carsten Willems, Thorsten Holz, and Konrad Rieck.
Technical report \s-1TR\-2009\-07,\s0 University of Mannheim, 2009
.PP
The input of \fBmalheur\fR is a \fIdataset\fR containing reports of malware
behavior. The \fIdataset\fR is provided either as a directory or a
compressed archive containing the reports. \fBmalheur\fR supports the
following formats for compressed archives: tar.gz, zip, pax and cpio.
A \fIreport\fR is a textual document describing the recorded activity of
a malware program, where individual events are separated by delimiter
characters, such as white space or carriage return.  The events in a
report are expected to be in sequential order if n\-grams with n\ >\ 1 are extracted. If the behavior is represented using the
malware instruction set (\s-1MIST\s0) further options may be selected.
.SH "ACTIONS"
.IX Header "ACTIONS"
\&\fBmalheur\fR supported different \fIactions\fR for analysis of a
\&\fIdataset\fR.  For all actions the reports are first mapped to a
high-dimensional vector space, such that each report is represented as
a feature vector. Recorded events or n\-grams of events are reflected
in different dimensions and the dissimilarity of behavior can be
assessed geometrically by computing distances and angles.
.IP "\fBdistance\fR" 12
.IX Item "distance"
If this action is specified, \fBmalheur\fR computes a distance matrix for
\&\fIdataset\fR. The entry (i,j) of this matrix reflects the distance
(dissimilarity) of the reports i and j. The distance values lie in the
range 0 to \fIsqrt\fR\|(2). The distance matrix is written to \fIoutput_file\fR.
.IP "\fBprototype\fR" 12
.IX Item "prototype"
If this action is specified, \fBmalheur\fR determines a set of
\&\fIprototypes\fR representing \fIdataset\fR. The prototypes are selected
from the contained reports, such that the distance from any report to
its nearest prototype is minimized. The prototype assignment of
\&\fIdataset\fR is written to \fIoutput_file\fR.
.IP "\fBcluster\fR" 12
.IX Item "cluster"
If this action is specified, \fBmalheur\fR performs a \fIclustering\fR of
\&\fIdataset\fR. The clustering is first determined on prototypes and then
propagated to all reports. Small clusters with too few members are
merged in a \fIrejection cluster\fR. The prototypes representing accepted
clusters are stored as internal state of \fBmalheur\fR for later
classification and incremental analysis. The clustering of \fIdataset\fR
is written to \fIoutput_file\fR.
.IP "\fBclassify\fR" 12
.IX Item "classify"
If this action is specified, \fBmalheur\fR performs a \fIclassification\fR
of \fIdataset\fR. Each report is either assigned to the nearest prototype
of a known cluster or rejected as unknown. This action requires that a
clustering has been performed beforehand and an internal state of
\&\fBmalheur\fR exists. The classification of \fIdataset\fR is written to
\&\fIoutput_file\fR.

CHANGES
2012-12-27  Konrad Rieck <konrad@mlsec.org>

    - fixed annoying ptr warning

    - updated autotool config

    - changed order of headers

2012-05-26  Konrad Rieck <konrad@mlsec.org>

    - added md README

2012-02-05  Konrad Rieck <konrad@mlsec.org>

    - improved loading of state. started support for info

    - changed working dir of Malheur to local-state dir

    - Malheur development continues in Goettingen now

2012-01-22  Konrad Rieck <konrad@mlsec.org>

    - minor changes to env scripts

2011-09-23  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - sync

2011-08-25  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - rewrote default configuration code

2011-08-24  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - fixed another libconfig related bug.

    - fixed problem with FLT_NONE constant. it is not simply NAN

2011-07-09  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - sync

    - missing files

2011-07-04  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - fixed printing of configuration

2011-04-19  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - updated manual page

2011-04-12  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - missed one release, so back to previous version number

2011-03-29  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - added support for longopts

2010-12-09  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - some new todos

2010-10-08  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - updated documentaiton

    - support for computing prototype distances

2010-10-07  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - updated documentation and added changes file

    - improved output of shared n-grams

2010-09-16  Konrad Rieck <rieck@node16.ml.tu-berlin.de>

    - switched to libconfig 1.4.x

2010-08-06  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - updated man page

    - changed email address

2010-08-04  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - improved computation of shared n-grams

2010-08-03  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - support for shared ngrams

2010-02-16  Konrad Rieck <rieck@cs.tu-berlin.de>

    - fixed bug with non-deterministic selection of first prototype

2010-02-03  Konrad Rieck <rieck@cs.tu-berlin.de>

    - added support to disable saving

2010-01-16  Konrad Rieck <rieck@cs.tu-berlin.de>

    - general code clean-up (casting, printf formats, etc)

2010-01-13  Konrad Rieck <rieck@cs.tu-berlin.de>

    - added local uthash version as fallback

2010-01-11  Konrad Rieck <rieck@cs.tu-berlin.de>

    - removed internal uthash

2010-01-01  Konrad Rieck <rieck@cs.tu-berlin.de>

    - fixed bug in classification mode

2009-12-28  Konrad Rieck <rieck@cs.tu-berlin.de>

    - improved documentation and examples

2009-12-16  Konrad Rieck <rieck@cs.tu-berlin.de>

    - fixed installation problems

2009-12-12  Konrad Rieck <rieck@cs.tu-berlin.de>

    - minor fixes to math. normalized vectors are mandatory

    - changed MIST format to latest

    - minor fixes

2009-12-05  Konrad Rieck <rieck@cs.tu-berlin.de>

    - added documenation for hash_seed parameters

2009-12-05  Konrad Rieck <rieck@fluffy.my.domain>

    - support changing the seed of MD5 in feature extraction

    - fixes for OpenBSD (including annoying FLT_NONE hack)

    - fixed minimalistic  matlab interface

2009-11-07  Konrad Rieck <rieck@cs.tu-berlin.de>

    - extension and spell fixes to manual

2009-11-07  Konrad Rieck <rieck@node04.ml.tu-berlin.de>

    - openmp enabled by default

2009-11-06  Konrad Rieck <rieck@cs.tu-berlin.de>

    - improved manpage egenration

2009-10-30  Konrad Rieck <rieck@cs.tu-berlin.de>

    - fixes for distribution

    - support to write version in latex

    - added nice output for pod file

    - added pod infavor of man.in

    - fixes to documentation

    - changed creation of man page

    - first paragraph

    - fixed copyright notice

    - minor fixes

2009-10-16  Konrad Rieck <rieck@cs.tu-berlin.de>

    - several improvements of doxygen documentation

2009-10-15  Konrad Rieck <rieck@cs.tu-berlin.de>

    - improved layout wrt assignments vs. classification using prototyes

    - changed file names

    - improved run-time of prototype extraction

    - minor speed improvements

2009-10-14  Konrad Rieck <rieck@cs.tu-berlin.de>

    - linux and matlab fixes

    - improved incremental output

    - first version of incremental analysis

2009-10-13  Konrad Rieck <rieck@cs.tu-berlin.de>

    - long overdue code formating

    - fix to automake

    - minor fixes

    - added classification test case

    - added clustering test case

    - added test case for prototype extraction

    - minor fixes in code layout


src/malheur.c
/*
 * MALHEUR - Automatic Analysis of Malware Behavior
 * Copyright (c) 2009-2015 Konrad Rieck (konrad@mlsec.org)
 * University of Goettingen, Berlin Institute of Technology
 * --
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 3 of the License, or (at your
 * option) any later version.  This program is distributed without any
 * warranty. See the GNU General Public License for more details.
 * --
 */

#include "config.h"
#include "malheur.h"
#include "common.h"
#include "mconfig.h"
#include "ftable.h"
#include "fmath.h"
#include "proto.h"
#include "util.h"
#include "cluster.h"
#include "class.h"
#include "export.h"

/* Global variables */
int verbose = 0;
config_t cfg;

static int print_conf = FALSE;
static char **input_files = NULL;
static int input_len = 0;
static int reset = FALSE;
static int save = TRUE;
static char *fvec_dump = NULL;
static malheur_action_t action = PROTOTYPE;
static malheur_state_t mstate;

/* Option string */
#define OPTSTRING       "no:c:s:hvVDC"

/**
 * Array of options of getopt_long()
 */
static struct option longopts[] = {
    {"reset", 0, NULL, 1001},
    {"dry", 0, NULL, 'n'},
    {"verbose", 0, NULL, 'v'},
    {"version", 0, NULL, 'V'},
    {"help", 0, NULL, 'h'},
    {"fvec_dump", 1, NULL, 1002},
    {"print_config", 0, NULL, 'C'},
    {"print_defaults", 0, NULL, 'D'},
    {"help", 0, NULL, 'h'},
    {"output_file", 1, NULL, 'o'},
    {"state_dir", 1, NULL, 's'},
    {NULL}
};

/**
 * Print usage of command line tool
 * @param argc Number of arguments
 * @param argv Argument values
 */
static void print_usage(void)
{
    printf("Usage: malheur [options] <action> <dataset>\n"
           "\nActions:\n"
           "  distance       Compute distance matrix for malware reports\n"
           "  prototype      Extract prototypes from malware reports\n"
           "  protodist      Compute distance matrix for prototypes\n"
           "  cluster        Cluster malware reports into similar groups\n"
           "  classify       Classify malware reports using labeled prototypes\n"
           "  increment      Incremental analysis of malware reports\n"
           "  info           Print information about internal state of Malheur\n"
           "\nOptions:\n"
           "  -s,  --state_dir <dir>       Set directory for internal state.\n"
           "  -o,  --output_file <file>    Set output file for analysis results.\n"
           "  -c,  --config_file <file>    Set configuration file.\n"
           "  -n,  --dry                   Dry run. Don't change internal state.\n"
           "       --reset                 Reset internal state of Malheur.\n"
           "       --fvec_dump <file>      Dump feature vectores in LibSVM format.\n"
           "  -C,  --print_config          Print the current configuration.\n"
           "  -D,  --print_defaults        Print the default configuration.\n"
           "  -v,  --verbose               Increase verbosity.\n"
           "  -V,  --version               Print version and copyright.\n"
           "  -h,  --help                  Print this help screen.\n"
           "\nSee manual page malheur(1) for more information.\n");
}

/**
 * Print configuration
 * @param msg Text to add to output
 */
static void print_config(char *msg)
{
    malheur_version(stderr);
    fprintf(stderr, "# ---\n# %s\n", msg);
    config_fprint(stderr, &cfg);
}


/**
 * Parse command line options
 * @param argc Number of arguments
 * @param argv Argument values
 */
static void malheur_parse_options(int argc, char **argv)
{
    int ch;

    /* reset getopt */
    optind = 0;

    while ((ch = getopt_long(argc, argv, OPTSTRING, longopts, NULL)) != -1) {
        switch (ch) {
        case 'c':
            /* Empty. See malheur_load_config() */
            break;
        case 'n':
            save = FALSE;
            break;
        case 1001:
            reset = TRUE;
            break;
        case 'v':
            verbose++;
            break;
        case 's':
            config_set_string(&cfg, "generic.state_dir", optarg);
            break;
        case 'o':
            config_set_string(&cfg, "generic.output_file", optarg);
            break;
        case 1002:
            fvec_dump = optarg;
            break;
        case 'V':
            malheur_version(stdout);
            exit(EXIT_SUCCESS);
            break;
        case 'D':
            print_config("Default configuration");
            exit(EXIT_SUCCESS);
            break;
        case 'C':
            print_conf = TRUE;
            break;
        case 'h':
        case '?':
            print_usage();
            exit(EXIT_SUCCESS);
            break;
        }
    }

    /* Check configuration */
    if (!config_check(&cfg)) {
        exit(EXIT_FAILURE);
    }

    /* We are through with parsing. Print the config if requested */
    if (print_conf) {
        print_config("Current configuration");
        exit(EXIT_SUCCESS);
    }

    argc -= optind;
    argv += optind;

    if (argc < 1)
        fatal("the <action> argument is required");

    /* Argument: action */
    if (!strcasecmp(argv[0], "prototype")) {
        action = PROTOTYPE;
    } else if (!strcasecmp(argv[0], "distance")) {
        action = DISTANCE;
    } else if (!strcasecmp(argv[0], "cluster")) {
        action = CLUSTER;
    } else if (!strcasecmp(argv[0], "classify")) {
        action = CLASSIFY;
    } else if (!strcasecmp(argv[0], "increment")) {
        action = INCREMENT;
    } else if (!strcasecmp(argv[0], "protodist")) {
        action = PROTODIST;
    } else if (!strcasecmp(argv[0], "info")) {
        action = INFO;
    } else {
        fatal("Unknown analysis action '%s'", argv[0]);
    }


    if (argc < 2 && action != PROTODIST && action != INFO)
        fatal("the <dataset> argument is required");

    /* Assign input files */
    input_files = argv + 1;
    input_len = argc - 1;
}

/**
 * Load configuration
 * @param argc Number of arguments
 * @param argv Argument values
 */
static void malheur_load_config(int argc, char **argv)
{
    char *cfg_file = NULL;
    int ch;

    /* Check for config file in command line */
    while ((ch = getopt_long(argc, argv, OPTSTRING, longopts, NULL)) != -1) {
        switch (ch) {
        case 'c':
            cfg_file = optarg;
            break;
        case 'h':
        case '?':
            print_usage();
            exit(EXIT_SUCCESS);
            break;
        default:
            /* empty */
            break;
        }
    }

    /* Init and load configuration */
    config_init(&cfg);

    if (cfg_file != NULL) {
        if (config_read_file(&cfg, cfg_file) != CONFIG_TRUE)
            fatal("Could not read configuration (%s in line %d)",
                  config_error_text(&cfg), config_error_line(&cfg));
    }

    /* Check configuration and set defaults */
    if (!config_check(&cfg)) {
        exit(EXIT_FAILURE);
    }

}


/**
 * Initialize malheur tool
 */
static void malheur_init()
{
    const char *state_dir;
    double shared;

    /* Init feature lookup table */
    config_lookup_float(&cfg, "cluster.shared_ngrams", &shared);
    if (shared > 0.0) {
        ftable_init();
    }

    /* Reset internal state */
    memset(&mstate, 0, sizeof(mstate));
    config_lookup_string(&cfg, "generic.state_dir", &state_dir);

    /* Create directory if not available */
    if (access(state_dir, W_OK)) {
        if (verbose > 0)
            printf("Creating state directory '%s'\n", state_dir);
        mkdir(state_dir, 0700);
    }

    /* Print configuration */
    if (verbose > 1)
        config_print(&cfg);
}


/**
 * Loads data from archives/directories into feature vectors
 * @return array of feature vectors
 */
static farray_t *malheur_load()
{
    farray_t *fa = NULL;
    for (int i = 0; i < input_len; i++) {
        /* Argument: Input */
        if (access(input_files[i], R_OK)) {
            warning("Could not access '%s'", input_files[i]);
            continue;
        }

        farray_t *f = farray_extract(input_files[i]);
        if (f) {
            fa = farray_merge(fa, f);
        }
    }

    if (!fa)
        fatal("No data available.");

    /* Dump feature vectors to file */
    if (fvec_dump)
        farray_save_libsvm_file(fa, fvec_dump);

    return fa;
}

/**
 * Saves the internal Malheur state. The state is used during incremental
 * analysis to distinguig clusters obtained during different runs
 * @param run Current run of analysis
 * @param proto Number of prototypes
 * @param rej Number of rejected reports
 */
static void malheur_save_state()
{
    FILE *f;
    const char *state_dir;
    char state_file[512];

    config_lookup_string(&cfg, "generic.state_dir", &state_dir);
    snprintf(state_file, 512, "%s/%s", state_dir, STATE_FILE);

    if (verbose > 0)
        printf("Saving internal state to '%s'.\n", state_file);

    f = fopen(state_file, "w");
    if (!f) {
        error("Could not open state file '%s'.", state_file);
        return;
    }

    fprintf(f, "run = %u\nprototypes = %u\nrejected = %u\n",
            mstate.run, mstate.num_proto, mstate.num_reject);

    fclose(f);
}

/**
 * Loads the internal Malheur state. The state is used during incremental
 * analysis to distinguig clusters obtained during different runs
 */
static int malheur_load_state()
{
    const char *state_dir;
    char state_file[512];
    FILE *f;
    int ret;

    config_lookup_string(&cfg, "generic.state_dir", &state_dir);
    snprintf(state_file, 512, "%s/%s", state_dir, STATE_FILE);

    if (access(state_file, R_OK))
        return FALSE;

    if (verbose > 0)
        printf("Loading internal state to '%s'.\n", state_file);

    f = fopen(state_file, "r");
    if (!f) {
        error("Could not open state file '%s'.", state_file);
        return FALSE;
    }

    ret = fscanf(f, "run = %u\nprototypes = %u\nrejected = %u\n",
                 &mstate.run, &mstate.num_proto, &mstate.num_reject);

    if (ret != 3) {
        error("Could not parse state file '%s'.", state_file);
        return FALSE;
    }

    fclose(f);
    return TRUE;
}

/**
 * Determines prototypes for the given malware reports
 */
static void malheur_prototype()
{
    assign_t *as;
    farray_t *fa, *pr;
    const char *state_dir, *output_file;
    char proto_file[512];

    config_lookup_string(&cfg, "generic.output_file", &output_file);
    config_lookup_string(&cfg, "generic.state_dir", &state_dir);
    snprintf(proto_file, 512, "%s/%s", state_dir, PROTO_FILE);

    /* Load data */
    fa = malheur_load();

    /* Extract prototypes */
    pr = proto_extract(fa, &as);
    if (verbose > 1)
        farray_print(pr);

    /* Save prototypes */
    if (save)
        farray_save_file(pr, proto_file);

    /* Export prototypes */
    export_proto(pr, fa, as, output_file);

    /* Clean up */
    assign_destroy(as);
    farray_destroy(pr);
    farray_destroy(fa);
}

/**
 * Clusters the given malware reports
 */
static void malheur_cluster()
{
    assign_t *as;
    farray_t *fa, *pr, *pn, *re;
    const char *state_dir, *output_file;
    char proto_file[512], reject_file[512];

    config_lookup_string(&cfg, "generic.output_file", &output_file);
    config_lookup_string(&cfg, "generic.state_dir", &state_dir);
    snprintf(proto_file, 512, "%s/%s", state_dir, PROTO_FILE);
    snprintf(reject_file, 512, "%s/%s", state_dir, REJECT_FILE);

    /* Load data */
    fa = malheur_load();

    /* Extract prototypes */
    pr = proto_extract(fa, &as);

    /* Cluster prototypes and extrapolate */
    cluster_t *c = cluster_linkage(pr, 0);
    cluster_extrapolate(c, as);
    cluster_trim(c);

    /* Save prototypes */
    pn = cluster_get_prototypes(c, as, pr);
    if (save)
        farray_save_file(pn, proto_file);
    farray_destroy(pn);

    /* Save rejected feature vectors */
    re = cluster_get_rejected(c, fa);
    if (save)
        farray_save_file(re, reject_file);
    farray_destroy(re);

    /* Export clustering */
    export_cluster(c, pr, fa, as, output_file);

    /* Export shared n-grams */
    export_shared_ngrams(c, fa, output_file);

    /* Clean up */
    cluster_destroy(c);
    assign_destroy(as);
    farray_destroy(pr);
    farray_destroy(fa);
}

/**
 * Classify the given malware reports
 */
static void malheur_classify()
{
    assign_t *as;
    farray_t *fa, *pr, *re;
    const char *state_dir, *output_file;
    char proto_file[512], reject_file[512];

    config_lookup_string(&cfg, "generic.output_file", &output_file);
    config_lookup_string(&cfg, "generic.state_dir", &state_dir);
    snprintf(proto_file, 512, "%s/%s", state_dir, PROTO_FILE);
    snprintf(reject_file, 512, "%s/%s", state_dir, REJECT_FILE);

    /* Check for prototype file */
    if (access(proto_file, R_OK))
        fatal("No prototype file for classifcation available");

    /* Load data */
    fa = malheur_load();

    /* Load prototypes */
    pr = farray_load_file(proto_file);

    /* Apply classification */
    as = class_assign(fa, pr);

    /* Save rejected feature vectors */
    re = class_get_rejected(as, fa);
    if (save)
        farray_save_file(re, reject_file);
    farray_destroy(re);

    /* Export classification */
    export_class(pr, fa, as, output_file);

    /* Clean up */
    assign_destroy(as);
    farray_destroy(pr);
    farray_destroy(fa);
}

/**
 * Classify the given malware reports
 */
static void malheur_increment()
{
    farray_t *pr = NULL, *tmp, *pn, *re;
    assign_t *as;
    const char *state_dir, *output_file;
    char proto_file[512], reject_file[512];

    config_lookup_string(&cfg, "generic.output_file", &output_file);
    config_lookup_string(&cfg, "generic.state_dir", &state_dir);
    snprintf(proto_file, 512, "%s/%s", state_dir, PROTO_FILE);
    snprintf(reject_file, 512, "%s/%s", state_dir, REJECT_FILE);

    /* Load internal state */
    malheur_load_state();

    /* Load data including rejected stuff */
    farray_t *fa = malheur_load();
    if (!access(reject_file, F_OK)) {
        tmp = farray_load_file(reject_file);
        fa = farray_merge(fa, tmp);
    }

    /* Classification */
    if (!access(proto_file, R_OK)) {
        pr = farray_load_file(proto_file);

        /* Apply classification */
        as = class_assign(fa, pr);
        tmp = class_get_rejected(as, fa);

        /* Export results */
        export_increment1(pr, fa, as, output_file);

        /* Clean up */
        farray_destroy(fa);
        farray_destroy(pr);
        assign_destroy(as);
        fa = tmp;
    } else {
        /* Export results */
        export_increment1(pr, fa, as, output_file);
    }

    /* Extract prototypes */
    pr = proto_extract(fa, &as);

    /* Cluster prototypes and extrapolate */
    cluster_t *c = cluster_linkage(pr, mstate.run + 1);
    cluster_extrapolate(c, as);
    cluster_trim(c);

    /* Save prototypes vectors */
    pn = cluster_get_prototypes(c, as, pr);
    if (save)
        farray_append_file(pn, proto_file);

    /* Save rejeted feature vectors */
    re = cluster_get_rejected(c, fa);
    if (save)
        farray_save_file(re, reject_file);

    /* Update state */
    mstate.run++;
    mstate.num_proto = pn->len;
    mstate.num_reject = re->len;

    /* Save state */
    if (save)
        malheur_save_state();

    /* Export results */
    export_increment2(c, pr, fa, as, output_file);

    /* Clean up */
    cluster_destroy(c);
    assign_destroy(as);

    farray_destroy(re);
    farray_destroy(pn);
    farray_destroy(pr);
    farray_destroy(fa);
}

/**
 * Display information about internal state of Malheur
 */
static void malheur_info()
{
    const char *state_dir;
    char state_file[512];

    config_lookup_string(&cfg, "generic.state_dir", &state_dir);
    snprintf(state_file, 512, "%s/%s", state_dir, STATE_FILE);

    /* Load internal state */
    if (!malheur_load_state()) {
        printf("No internal state stored in %s\n", state_dir);
        return;
    }

    printf("Internal state from %s\n", state_file);
    printf("       Malheur run: %u\n", mstate.run);
    printf(" Stored prototypes: %u\n", mstate.num_proto);
    printf("  Rejected reports: %u\n", mstate.num_reject);

}

/**
 * Computes a distance matrix and saves the result to a file
 */
static void malheur_distance()
{
    const char *output_file;
    config_lookup_string(&cfg, "generic.output_file", &output_file);

    /* Load data */
    farray_t *fa = malheur_load();

    /* Allocate distance matrix */
    double *d = malloc(fa->len * fa->len * sizeof(double));
    if (!d)
        fatal("Could not allocate similarity matrix");

    /* Compute distance matrix */
    farray_dist(fa, fa, d);

    /* Save distance matrix */
    export_dist(d, fa, output_file);

    /* Clean up */
    free(d);
    farray_destroy(fa);
}

/**
 * Computes a distance matrix for the prototypes and saves the result to a file
 */
static void malheur_protodist()
{
    farray_t *pr;
    const char *state_dir, *output_file;
    char proto_file[512];

    config_lookup_string(&cfg, "generic.output_file", &output_file);
    config_lookup_string(&cfg, "generic.state_dir", &state_dir);
    snprintf(proto_file, 512, "%s/%s", state_dir, PROTO_FILE);

    /* Check for prototype file */
    if (access(proto_file, R_OK))
        fatal("No prototype file for classifcation available");

    /* Load prototypes */
    pr = farray_load_file(proto_file);
    if (verbose > 1)
        farray_print(pr);

    /* Allocate distance matrix */
    double *d = malloc(pr->len * pr->len * sizeof(double));
    if (!d)
        fatal("Could not allocate similarity matrix");

    /* Compute distance matrix */
    farray_dist(pr, pr, d);

    /* Save distance matrix */
    export_dist(d, pr, output_file);

    /* Clean up */
    free(d);
    farray_destroy(pr);
}

/**
 * Exits the malheur tool.
 */
static void malheur_exit()
{
    /* Destroy feature lookup table */
    if (verbose > 0)
        ftable_print();
    ftable_destroy();

    /* Destroy configuration */
    config_destroy(&cfg);
}

/**
 * Main function of Malheur
 * @param argc Number of arguments
 * @param argv Argument values
 * @return Exit code
 */
int main(int argc, char **argv)
{

    malheur_load_config(argc, argv);
    malheur_parse_options(argc, argv);
    malheur_init();

    /* Perform action */
    switch (action) {
    case DISTANCE:
        malheur_distance();
        break;
    case PROTOTYPE:
        malheur_prototype();
        break;
    case CLUSTER:
        malheur_cluster();
        break;
    case CLASSIFY:
        malheur_classify();
        break;
    case INCREMENT:
        malheur_increment();
        break;
    case PROTODIST:
        malheur_protodist();
        break;
    case INFO:
        malheur_info();
    }

    malheur_exit();
    return EXIT_SUCCESS;
}
